akka-typed/src/main/scala/akka/typed/internal/DispatchersImpl.scala

```diff
@@ -5,32 +5,223 @@ package akka.typed
+        // That shouldn't happen often and in case it does the actual ExecutorService isn't
+        // created until used, i.e. cheap.
+        val newConfigurator =
+          if (cachingConfig.hasPath(id)) configuratorFrom(config(id))
+          else throw new ConfigurationException(s"Dispatcher [$id] not configured")
+
+        dispatcherConfigurators.putIfAbsent(id, newConfigurator) match {
+          case null     ⇒ newConfigurator
+          case existing ⇒ existing
+        }
+
+      case existing ⇒ existing
+    }
+  }
+
+  /**
+   * INTERNAL API
+   */
+  private[akka] def config(id: String): Config = {
+    import scala.collection.JavaConverters._
+    def simpleName = id.substring(id.lastIndexOf('.') + 1)
+    idConfig(id)
+      .withFallback(settings.config.getConfig(id))
+      .withFallback(ConfigFactory.parseMap(Map("name" → simpleName).asJava))
+      .withFallback(defaultDispatcherConfig)
+  }
+
+  private def idConfig(id: String): Config = {
+    import scala.collection.JavaConverters._
+    ConfigFactory.parseMap(Map("id" → id).asJava)
+  }
+
+  /**
+   * INTERNAL API
+   *
+   * Creates a MessageDispatcherConfigurator from a Config.
+   *
+   * The Config must also contain a `id` property, which is the identifier of the dispatcher.
+   *
+   * Throws: IllegalArgumentException if the value of "type" is not valid
+   *         IllegalArgumentException if it cannot create the MessageDispatcherConfigurator
+   */
+  private def configuratorFrom(cfg: Config): MessageDispatcherConfigurator = {
+    if (!cfg.hasPath("id")) throw new ConfigurationException("Missing dispatcher 'id' property in config: " + cfg.root.render)
+
+    cfg.getString("type") match {
+      case "Dispatcher"       ⇒ new DispatcherConfigurator(cfg, prerequisites)
+      case "PinnedDispatcher" ⇒ new PinnedDispatcherConfigurator(cfg, prerequisites)
+      case fqn ⇒
+        val args = List(classOf[Config] → cfg)
+        prerequisites.dynamicAccess.createInstanceFor[DispatcherConfigurator](fqn, args).recover({
+          case exception ⇒
+            throw new ConfigurationException(
+              ("Cannot instantiate DispatcherConfigurator type [%s], defined in [%s], " +
+                "make sure it has constructor with [com.typesafe.config.Config] and " +
+                "[akka.dispatch.DispatcherPrerequisites] parameters")
+                .format(fqn, cfg.getString("id")), exception)
+        }).get
+    }
+  }
+}
+
+/**
+ * Base class to be used for hooking in new dispatchers into Dispatchers.
+ */
+abstract class MessageDispatcherConfigurator(_config: Config, val prerequisites: d.DispatcherPrerequisites) {
+
+  val config: Config = new d.CachingConfig(_config)
+
+  /**
+   * Returns an instance of MessageDispatcher given the configuration.
+   * Depending on the needs the implementation may return a new instance for
+   * each invocation or return the same instance every time.
+   */
+  def dispatcher(): ExecutionContextExecutor
```

Comment:
This is the only and big change from `akka.dispatch.MessageDispatcherConfigurator`: the whole notion of a MessageDispatcher is gone and replaced by a bare `ExecutionContextExecutor`. Reimplementing the _absolutely baroque_ configuration semantics of the old system did not seem like a good idea, hence I reuse all that stuff and only cut out the MessageDispatcher middleman.